themes/default/content/blog/csharp-static-analyzer/index.md

```diff
@@ -0,0 +1,130 @@
+meta_desc: "Introducing a static-code analyzer for C# which detects programmer errors at compile-time"
+meta_image: enhanced-static-code-analysis.png
+date: "2022-07-02"
+---
+
+When I started using Pulumi for the first time, I used C# as my language of choice for defining infrastructure. I start by creating resources and provide their parameters through argument objects. My code would look something like the following:
+
+```cs
+var resourceGroup = new ResourceGroup("resourceGroup");
+var storageAccount = new StorageAccount("storageAccount", new StorageAccountArgs
+{
+    ResourceGroupName = resourceGroup.Name,
+    Sku = new SkuArgs { Name = SkuName.Standard_LRS }
+});
+```
+
+Here, I am creating a resource group in Azure, then creating a storage account that belongs to that resource group. Pretty simple stuff. The IDE tells me that the code is good and doesn't report any error but just to be sure everything is as it should be, I build the project and the code compiles successfully. Now I am confident and ready to deploy the infrastructure.
+
+I run `pulumi preview` and to my surprise, it fails. The C# program throws an exception at runtime:
+
+```
+Diagnostics:
+  pulumi:pulumi:Stack (AzureExample-dev):
+    error: Running program '/Users/zaid/projects/AzureExample/bin/Debug/net6.0/AzureExample.dll' failed with an unhandled exception:
+    System.ArgumentNullException: [Input] Pulumi.AzureNative.Storage.StorageAccountArgs.Kind is required but was not given a value (Parameter 'Kind')
+```
+
+What happened? It turns out I didn't provide a required property `Kind` when instantiating the type `StorageAccountArgs` for the `StorageAccount`.
+
+This is an unfortunate result because I had expected that when my  program compiled, that it was valid and ready to deploy.
+
+Even though the fix for this case is simple:
+
+```cs
+var storageAccount = new StorageAccount("storageAccount", new StorageAccountArgs
+{
+    ResourceGroupName = resourceGroup.Name,
+    Sku = new SkuArgs { Name = SkuName.Standard_LRS },
+    // adding this property
+    Kind = Kind.StorageV2
+});
+```
+
+There is a bigger problem in the API of the resource creation which is that it doesn't tell you that you missed a required property at compile-time. The underlying cause is because we are using property initializers for `StorageAccountArgs` with a _parameterless_ contructor. This is the case when creating _any_ resource across the cloud providers available in the Pulumi ecosystem.
+
+What can we do to fix the problem?
+
+First solution that comes to mind is to fix the generated .NET API for each cloud provider available with Pulumi: adding the required parameters as part of the contructor of arguments type and removing the _parameterless_ contructor.
+
+This however means breaking each and every existing C# program running with Pulumi since the removal of the empty constructor. Here at Pulumi, we want to ensure that changes to our APIs are backward-compatible so this is a no-go.
```

Comment:
This is not the only problem. We can't use constructors because constructor arguments are positional. This means that we'd have to guarantee the order of required arguments, and there's no natural ordering other than alphabetical (which likely looks quite confusing).

Reply:
I mentioned issue https://github.com/pulumi/pulumi/issues/3808 so people can read more about the underlying issue there so I didn't reiterate it here